variables.tf

```diff
@@ -1,11 +1,153 @@
+  type        = list(string)
+}
+
+variable "associated_subnets" {
+  type        = list(string)
+  description = "List of subnets to associate with the VPN endpoint"
+}
+
+variable "authorization_rules" {
+  type = list(object({
+    name                 = string
+    access_group_id      = string
+    authorize_all_groups = bool
+    description          = string
+    target_network_cidr  = string
+  }))
+  description = "List of objects describing the authorization rules for the client vpn"
+}
+
+variable "client_cidr_block" {
+  type        = string
+  description = "(optional) describe your variable"
+}
+
+variable "cloudwatch_log_retention_days" {
+  default     = 30
+  description = "How long to keep VPN logs. Possible values are: 1, 3, 5, 7, 14, 30, 60, 90, 120, 150, 180, 365, 400, 545, 731, 1827, 3653, and 0. If you select 0, the events in the log group are always retained and never expire."
+  type        = number
+
+  validation {
+    error_message = "Invalid value. Possible values are: 1, 3, 5, 7, 14, 30, 60, 90, 120, 150, 180, 365, 400, 545, 731, 1827, 3653, and 0. If you select 0, the events in the log group are always retained and never expire."
+
+    condition = (
+      contains(
+        [
+          1,
+          3,
+          5,
+          7,
+          14,
+          30,
+          60,
+          90,
+          120,
+          150,
+          180,
+          365,
+          400,
+          545,
+          731,
+          1827,
+          3653,
+          0
+        ],
+        var.cloudwatch_log_retention_days
+      )
+    )
+  }
+}
 
 variable "name" {
-  description = "Moniker to apply to all resources in the module"
+  description = "Name to associate with various resources"
   type        = string
 }
 
+variable "saml_metadata_document" {
+  default     = null
+  description = "Optional SAML metadata document. Must include this or `saml_provider_arn`"
+  type        = string
+}
+
+variable "saml_provider_arn" {
+  default     = null
+  description = "Optional SAML provider ARN. Must include this or `saml_metadata_document`"
+  type        = string
+
+  validation {
+    error_message = "Invalid SAML provider ARN."
+
+    condition = (
+      var.saml_provider_arn == null ||
+      try(length(regexall(
+        "^arn:aws:iam::(?P<account_id>\\d{12}):saml-provider/(?P<provider_name>[\\w+=,\\.@-]+)$",
+        var.saml_provider_arn
+        )) > 0,
+        false
+    ))
+  }
+}
+
+module "saml_is_defined" {
+  source  = "rhythmictech/errorcheck/terraform"
+  version = "~> 1.0"
+
+  assert        = var.saml_metadata_document != null || var.saml_provider_arn != null
+  error_message = "Must define a value for either `saml_metadata_document` or `saml_provider_arn`."
+}
+
+module "saml_not_defined_twice" {
+  source  = "rhythmictech/errorcheck/terraform"
+  version = "~> 1.0"
+
+  assert        = ! (var.saml_metadata_document != null && var.saml_provider_arn != null)
```

Comment:
```suggestion
  assert        = var.saml_metadata_document == null || var.saml_provider_arn == null)
```

Reply:
Nope. I need an XOR, not an OR. XOR isn't available directly in Terraform, so I have to do this stupid shit.